themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md

```diff
@@ -0,0 +1,89 @@
+
+Using stack references, the Zephyr team can make the necessary connections they needed:
+
+* The code handling the Kubernetes cluster can reference the VPC ID and subnet IDs from the base infrastructure layer.
+* Similarly, the code that is responsible for deploying the online store services can reference the necessary Kubernetes connection details (the Kubeconfig) for the Kubernetes cluster.
+
+It's worthwhile to note that any information that needs to be accessible from another stack via a stack reference must be exported as a stack output in the source stack. If you don't mark it as a stack output, then it can't be used in a stack reference. Adding stack outputs after the fact requires little effort and has no impact on existing infrastructure.
+
+Also, any value retrieved via a stack reference is treated as a Pulumi Output, and therefore may require some extra work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps considerably in this situation. Some SDKs also have language-specific mechanisms that can help; for example, using Go's `.AsStringArrayOutput()` method on a `StackReference.GetOutput` statement makes referencing subnet IDs from other project much easier.
+
+While stack references are conceptually straightforward and not difficult to implement, there are some recommended practices to be mindful of regarding the use of stack outputs and stack references:
+
+1. When it comes to stack outputs, **export what is needed.** If it needs to be accessed from outside the stack, export it; otherwise, don't. It's simple to add more stack outputs after the fact, and as has been mentioned already this is done with a quick `pulumi up` that has no affect on existing infrastructure (you're modifying the stack object itself).
+2. **Be judicious in the use of stack references.** Each stack reference is a call to the backend to read a value, which adds some level of latency. Using lots of stack references (think more than 20 or so) can introduce notable latency in Pulumi operations.
+3. **If you need lots of stack references, use a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in a single call to the backend. Be aware, though, that you'll need to write the necessary code to understand/import/unmarshall that JSON object in the referring stack, so there is a small amount of extra work required in this situation.
```

Comment:
I might tweak this slightly to clarify we're talking about the exporting, not the consuming (but up to you -- feel free to ignore if you don't think it adds anything):

```suggestion
3. **If you need to expose lots of outputs on a stack, consider exporting a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in a single call to the backend. Be aware, though, that you'll need to write the necessary code to understand/import/unmarshall that JSON object in the referring stack, so there is a small amount of extra work required in this situation.
```

It might also be worth mentioning that this approach can also make it a little easier to codify the API contract between stacks and their consumers (vs. that contract consisting of "a bunch of optional strings"). 

Relatedly, we might want to mention that outputs are expected to be primitive/serializable types -- string, bool, object, etc. Occasionally folks think they can export whole resource instances (as you would in a typical program) and then realize this doesn't quite work, and are confused. (The docs may also mention this, though.)

Reply:
In 17aacd6f I've reworded per your suggestion, but there is still an outstanding question as to whether it is the stack outputs that are "costly" or if it is the stack references reading those outputs that's "costly" (from a latency perspective). I've asked @ringods & @pierskarsenbarg to weigh in on that, as I'm not sure.

Reply:
I think this has been resolved (refer to Slack discussions in #project-zephyr and #questions).